src/main/java/mate/zorii/bookstore/service/impl/OrderServiceImpl.java

```diff
@@ -0,0 +1,115 @@
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+import lombok.RequiredArgsConstructor;
+import mate.zorii.bookstore.dto.order.CreateOrderRequestDto;
+import mate.zorii.bookstore.dto.order.OrderItemResponseDto;
+import mate.zorii.bookstore.dto.order.OrderResponseDto;
+import mate.zorii.bookstore.dto.order.UpdateOrderStatusDto;
+import mate.zorii.bookstore.mapper.OrderMapper;
+import mate.zorii.bookstore.model.Order;
+import mate.zorii.bookstore.model.OrderItem;
+import mate.zorii.bookstore.model.User;
+import mate.zorii.bookstore.repository.OrderItemRepository;
+import mate.zorii.bookstore.repository.OrderRepository;
+import mate.zorii.bookstore.service.OrderService;
+import mate.zorii.bookstore.service.ShoppingCartService;
+import org.springframework.data.domain.Page;
+import org.springframework.data.domain.Pageable;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+@Service
+@RequiredArgsConstructor
+public class OrderServiceImpl implements OrderService {
+    private final OrderRepository orderRepository;
+    private final OrderItemRepository orderItemRepository;
+    private final OrderMapper orderMapper;
+    private final ShoppingCartService shoppingCartService;
+
+    @Override
+    @Transactional
+    public List<OrderResponseDto> updateOrderStatus(Long orderId, UpdateOrderStatusDto updateDto) {
+        Order updatedOrder = orderRepository.findById(orderId)
+                .map(order -> {
+                    order.setStatus(updateDto.status());
+                    return order;
+                }).orElseThrow(() -> new EntityNotFoundException("No order with id " + orderId));
+        orderRepository.save(updatedOrder);
+        return orderRepository.findAll().stream()
+                .map(orderMapper::toOrderDto)
+                .toList();
+    }
+
+    @Override
+    public Page<OrderItemResponseDto> getAllOrderItems(Long orderId, Pageable pageable) {
+        return orderItemRepository.findAllByOrderId(orderId, pageable)
+                .map(orderMapper::toOrderItemDto);
+    }
+
+    @Override
+    public Page<OrderResponseDto> getAllOrders(Long authenticatedUserId, Pageable pageable) {
+        return orderRepository.findAllByUser_Id(authenticatedUserId, pageable)
+                .map(orderMapper::toOrderDto);
+    }
+
+    @Override
+    public OrderItemResponseDto getOrderItem(Long orderId, Long itemId) {
+        return orderItemRepository.findByIdAndOrder_Id(itemId, orderId)
+                .map(orderMapper::toOrderItemDto)
+                .orElseThrow(() -> new EntityNotFoundException(
+                        String.format("No order item by id %d for orderId %d", itemId, orderId)));
+    }
+
+    @Override
+    @Transactional
+    public List<OrderResponseDto> placeOrder(CreateOrderRequestDto createDto,
+                                             Long authenticatedUserId) {
+        Order order = new Order();
```

Comment:
```suggestion
        ShoppingCart cart = cartRepository.findByUserId(userId);
        if (cart.getCartItems().isEmpty()) {
            .....
        }
        Order order = orderMapper.cartToOrder(cart, orderDto.shippingAddress());
        cart.clearCart();
        return orderMapper.toOrderDto(orderRepository.save(order));
```

Where in OrderMapper class

```
@Mapping(target = "id", ignore = true)
@Mapping(target = "total", source = "cart.cartItems", qualifiedByName = "total")
@Mapping(target = "orderItems", source = "cart.cartItems")
Order cartToOrder(ShoppingCart cart, String shippingAddress);
```

and 

```
@Named("total")
    default BigDecimal getTotal(Set<CartItem> cartItems) {
        return cartItems.stream()
                .map(i -> i.getBook().getPrice().multiply(BigDecimal.valueOf(i.getQuantity())))
                .reduce(BigDecimal.ZERO, BigDecimal::add);
    }
```

and 
```
@AfterMapping
    default void updateOrder(@MappingTarget Order order) {
        order.getOrderItems().forEach(oi -> oi.setOrder(order));
    }
```

and in ShoppingCart class
```
public void clearCart() {
        cartItems.clear();
    }
```